### ISSUE TITLE
Upgrade Python and Tensorflow

### ISSUE THREAD

**Issue by aazuspan** [lemma-osu/naip-cnn#27]
We're currently running the `tensorflow:2.13.0` Docker image which uses Python 3.8, which reached end-of-life over a year ago. Since we rely on some cloud-based services like Earth Engine and Weights & Biases, there's a real risk that this workflow stops working if there are breaking changes to those services and we're unable to update packages. Unfortunately, upgrading to the newer `tensorflow:2.20.0` image with Python 3.11 isn't trivial for a few different reasons.

## Model migration

The biggest issue is model compatibility. Our models are all saved in the recommended `.keras` format, but this was broken multiple times between `2.13` and `2.20` (see [this](https://stackoverflow.com/questions/77353981/layer-conv2d-11-expected-2-variables-but-received-0-variables-during-loading) and [this](https://github.com/tensorflow/tensorflow/issues/62363)), and models will need to be migrated in order for them to continue working. I haven't found a definitive guide on how to do that (most advice online suggests you continue using an old version or retrain your models, neither of which are practical for our case), but I found the steps below based on trial and error, and I was able to produce identical predictions from a single migrated model in `2.20`.

EDIT: I created a [private repo](https://github.com/lemma-osu/naip-cnn-model-migration) that automates the steps below.

1. Load a `.keras` model in the `2.13.0` Docker container and save it back out as a legacy `.h5` model.
1. Load the `.h5` model in a `2.15.0` environment (manually including the `R2Score2D` loss class) and save it back out to a new `.h5` model. For example:
    ```python
    model_path = "models/brisk-rain-129-2.13.h5"
    model = tf.keras.models.load_model(model_path, custom_objects={
        "R2Score2D": R2Score2D
    })
    model.save("models/brisk-rain-129-2.15.h5")
    ```
1. Load the `.h5` model from `2.15.0` in `2.20.0` and save it back out *again* as a `.keras` model.

We'll need to run those steps for every logged model in W&B and re-upload the migrated models. W&B supports a file versioning system, so I think we can just upload them as a new version without losing the old version, but I need to confirm that. We also need to make sure that we can continue to use the old models until the Docker container can be reliably updated to use the new models, since I'm assuming they are not backwards compatible.

## Code migration

I haven't dug into this or tried to train any new models in the `2.20` container, but I believe we'll need to migrate some code due to the implicit upgrade from Keras 2 to Keras 3 in `tensorflow==2.16`.

See https://keras.io/guides/migrating_to_keras_3/

## Docker warnings

Running the `2.20` container gives us lots of new warnings due to incompatible versions or missing plugins (I thought Docker was designed to avoid problems like this). I'll add details if/when I figure out how to resolve these.

### Protobuf version

```
/usr/local/lib/python3.11/dist-packages/google/protobuf/runtime_version.py:98: UserWarning: Protobuf gencode version 5.28.3 is exactly one major version older than the runtime version 6.31.1 at tensorflow/core/framework/attr_value.proto. Please update the gencode to avoid compatibility violations in the next runtime release.
```

### Tensorflow IO plugins

```
/usr/local/lib/python3.11/dist-packages/tensorflow_io/python/ops/__init__.py:98: UserWarning: unable to load libtensorflow_io_plugins.so: unable to open file: libtensorflow_io_plugins.so, from paths: ['/usr/local/lib/python3.11/dist-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so']
caused by: ['/usr/local/lib/python3.11/dist-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so: undefined symbol: _ZN3tsl5mutex6unlockEv']
  warnings.warn(f"unable to load libtensorflow_io_plugins.so: {e}")
/usr/local/lib/python3.11/dist-packages/tensorflow_io/python/ops/__init__.py:104: UserWarning: file system plugins are not loaded: unable to open file: libtensorflow_io.so, from paths: ['/usr/local/lib/python3.11/dist-packages/tensorflow_io/python/ops/libtensorflow_io.so']
caused by: ['/usr/local/lib/python3.11/dist-packages/tensorflow_io/python/ops/libtensorflow_io.so: undefined symbol: _ZNK10tensorflow4data11DatasetBase8FinalizeEPNS_15OpKernelContextESt8functionIFN4absl12lts_202308028StatusOrIN3tsl4core11RefCountPtrIS1_EEEEvEE']
```

This is probably caused by Tensorflow IO being incompatible with newer TF versions. The last TFIO release was in mid 2024 and is only compatible up to `tensorflow==2.16`. We rely on TFIO for handling HDF5 training data, and I don't think there are any alternatives, but recent issues in that repo suggest the project is abandoned. I'm not sure how critical the warnings above are, or whether they can be safely ignored.

See https://github.com/tensorflow/io?tab=readme-ov-file#tensorflow-version-compatibility

### Missing GPU libraries

```
W0000 00:00:1765305540.920133    2263 gpu_device.cc:2342] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
```

I installed [NVIDIA Container Toolkit](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html#with-apt-ubuntu-debian) in WSL as [recommended by Tensorflow](https://www.tensorflow.org/install/docker#gpu_support), but still get this warning and show no GPU devices when running:

```
docker run --rm --gpus all tensorflow/tensorflow:2.20.0-gpu-jupyter \
    python3 -c "import tensorflow as tf; print(tf.config.list_physical_devices('GPU'))"
```

Interestingly, I do see the GPU when I run `nvidia-smi` in the same container, so the problem seems to be on the tensorflow side.

```
docker run --rm --gpus all tensorflow/tensorflow:2.20.0-gpu-jupyter nvidia-smi
```

Looking back at older Tensorflow images, `2.17.0` is the last version that correctly recognizes the GPU.

### Possible inconsistent Adam optimizer

```
/usr/local/lib/python3.11/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 38 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))
```

**Comment by grovduck** [comment]
@aazuspan, you're much deeper in the weeds than I can hope to be, but I did go ahead and try your above command in my Ubuntu distro (was this the right one?) and it returned the same output as you had.  But it sounds like this isn't an option at present due to reliance on `tensorflow-io`.  I know nothing about `webdataset` although it looks like, it too, hasn't been active for six months (although, perhaps, it could just be stable).  

I think you are saying that we have the option of running Python 3.11 / Tensorflow 2.16 through a docker image at present, correct?  That maintains compatibility with `tensorflow-io` and addresses migrating the models from the older `.keras` format?  Is this a worthwhile intermediate step or would you prefer getting to the latest TF even if it means moving to a different training data format?

I *do* agree that this `uv` setup is much easier than Docker.

Forgive me if I don't understand all the moving parts.

**Comment by aazuspan** [comment]
> I know nothing about webdataset although it looks like, it too, hasn't been active for six months (although, perhaps, it could just be stable).

Good point. I also did some tests converting our existing training data, and webdataset almost doubles the file size compared to HDF since it stores each sample as an individual file with some per-file overhead. I like the transparency of basically just using a tar of PNG and JSON files, but it might not be the best choice after all.

> I think you are saying that we have the option of running Python 3.11 / Tensorflow 2.16 through a docker image at present, correct?

Yes, thanks for distilling my rambling notes to the important bit! I'm pretty confident that if we migrate models and a little bit of code, we could be back up and running pretty quickly with the `tensorflow:2.16.2` image and Python 3.11. The only thing that I'm a little wary about is that the image ships a release-candidate version of Python 3.11, but I don't know that that's actually an issue.

>  Is this a worthwhile intermediate step or would you prefer getting to the latest TF even if it means moving to a different training data format?

I'm starting to lean towards an incremental update to `2.16.2`, since that gets us over the hump from Keras 2 to 3 which should keep this project operational for generating predictions or training new models for the foreseeable future, without sinking a ton of time into it. Let me know what you think or if you have a preference for how to proceed.
